x-pack/plugins/apm/public/components/app/Main/useDelayedVisibility/Delayed/index.ts

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+type Callback = (isVisible: boolean) => void;
+
+export class Delayed {
+  private displayedAt = 0;
+  private hideDelayMs: number;
+  private isVisible = false;
+  private minimumVisibleDuration: number;
+  private showDelayMs: number;
+  private timeoutId?: number;
+
+  constructor({
+    minimumVisibleDuration = 1000,
+    showDelayMs = 50,
+    hideDelayMs = 50
+  } = {}) {
+    this.minimumVisibleDuration = minimumVisibleDuration;
+    this.hideDelayMs = hideDelayMs;
+    this.showDelayMs = showDelayMs;
+  }
+
+  private onChangeCallback: Callback = () => null;
+
+  private updateState(isVisible: boolean) {
+    window.clearTimeout(this.timeoutId);
+    const ms = !isVisible
+      ? Math.max(
+          this.displayedAt + this.minimumVisibleDuration - Date.now(),
+          this.hideDelayMs
+        )
+      : this.showDelayMs;
```

Comment:
Are these backwards? If the item is already visible, you would want the "hideDelayMs" I think? Might be reading it wrong...

Reply:
I think maybe the terminology used here is confusing. 
We currently only call `updateState(true)` if `isVisible` is already `false`.]

Meaning inside `updateState`, when `isVisible=true` `this.isVisible=false`. Maybe `isVisible` should be renamed to `nextVisibility` so it ready like

```js
    const ms = !nextVisibility
      ? Math.max(
          this.displayedAt + this.minimumVisibleDuration - Date.now(),
          this.hideDelayMs
        )
      : this.showDelayMs;
```

And perhaps `updateState` should start with:

```
if(nextVisibility === this.isVisible) {
  return;
}
```